aten/src/ATen/native/mkldnn/xpu/detail/QMatmul.cpp

```diff
@@ -0,0 +1,274 @@
+#include <ATen/Tensor.h>
+#include <ATen/core/Tensor.h>
+
+#include <ATen/native/mkldnn/xpu/detail/Attr.h>
+#include <ATen/native/mkldnn/xpu/detail/oneDNNContext.h>
+
+#include <oneapi/dnnl/dnnl.hpp>
+
+namespace at::native::onednn {
+
+void quantized_matmul(
+    at::Tensor mat1, // act
+    double input_scale,
+    int64_t input_zero_point,
+    at::Tensor mat2, // weight
+    at::Tensor& weight_scales,
+    at::Tensor& weight_zero_points,
+    at::Tensor& b_raw,
+    at::Tensor result, // output
+    double output_scale,
+    int64_t output_zero_point,
+    std::optional<c10::ScalarType> output_dtype,
+    std::optional<at::Tensor> other, // extra input for binary-post-op
+    double other_scale,
+    int64_t other_zero_point,
+    const c10::string_view& binary_post_op,
+    double binary_alpha,
+    const c10::string_view& unary_post_op,
+    torch::List<std::optional<at::Scalar>>& unary_post_op_args,
+    c10::string_view unary_post_op_algorithm) {
+  // [Note] Quantized Matrix Multiplication at XPU
+  // The following code integrates oneDNN quantized gemm. The quantization
+  // config we support:
+  // activation: s8&u8; per tensor calibrated; symmetric&asymmetric
+  // weight: s8; per_tensor/per_channel calibrated; symmetric
+  bool m2_trans = true;
+  auto attr = Attr(output_scale, output_zero_point);
+  construct_attr_by_post_op(
+      binary_post_op,
+      binary_alpha,
+      input_scale,
+      input_zero_point,
+      unary_post_op,
+      unary_post_op_args,
+      unary_post_op_algorithm,
+      attr);
+
+  size_t dims = result.dim();
+  at::Device curDevice = at::Device(at::kXPU, c10::xpu::current_device());
```

Comment:
This code is based on an assumption: the current device is the same as the device of input tensors. Do we need to add a check anywhere?

Reply:
As discussed above, the output is now created using input device directly. We may not need add extra check in the integration code with the modification. Appreciation for your suggestions.

Reply:
Please refer to https://github.com/pytorch/pytorch/blob/de9d6a25d719c75dccb1d00d6f3acfe448f06b85/aten/src/ATen/native/cuda/int4mm.cu#L1102-L1105, use `c10::DeviceGuard` instead of `CUDAGuard` here. 
The same device check is needed as well.

Reply:
Device guard is intended for switching device cases, At present, INT8 is not needed consider this issue? I have added some comments here https://github.com/pytorch/pytorch/pull/133307#discussion_r1906511554. We may consider device switch cases in another topic PR, I think. At present, we could directly assume no device switch.

Reply:
device switch is a very common usage scenario...

Reply:
Yes, I agree that it is common. What I want to express is, I don't know whether int8 flow currently could handle cross device properly, as CPU is presumed to default device currently. Cross devices is a special issue to verify with whole picture. I'd rather we run into problem before we formerly have cross device support in the whole int8 stack.